Error message for question mark operator should message conversion

I've been working with a lot of people who are either new to Rust or new to its latest features (like the question mark operator) and I'm noticing a lot of confusion around a specific error message.

Suppose you have the following code:

```rust
#[derive(Debug)]
struct Error1;
#[derive(Debug)]
struct Error2;

fn foo() -> Result<(), Error1> {
    Err(Error1)
}

fn main() -> Result<(), Error2> {
    foo()?;
    
    Ok(())
}
```

This gives you the error:

```
error[E0277]: the trait bound `Error2: std::convert::From<Error1>` is not satisfied
  --> src/main.rs:11:5
   |
11 |     foo()?;
   |     ^^^^^^ the trait `std::convert::From<Error1>` is not implemented for `Error2`
   |
   = note: required by `std::convert::From::from`
```

However, if you look at the actual code, it isn't obvious where the `From` trait is even being used.

The [documentation for the `From` trait](https://doc.rust-lang.org/std/convert/trait.From.html) actually does mention this briefly, though most people I encounter don't go as far as looking that up. The error message is pretty daunting if you're new to the language.

I propose that we add something to the error message when we see that they are using the question mark:

```
help: the question mark operation (`?`) implicitly performs a conversion
      on the error value using the `From` trait. If no conversion is
      defined, you will see this error message.
```

This conveys exactly where the `From` trait is being used and why they may be running into this issue. You certainly don't have to use this exact wording. Feel free to use whatever fits. :)

For what it's worth, [the current output points at the question mark now](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=a940a39dbcaa3175853c05e358040625):

```
error[E0277]: `?` couldn't convert the error to `Error2`
  --> src/main.rs:11:10
   |
11 |     foo()?;
   |          ^ the trait `std::convert::From<Error1>` is not implemented for `Error2`
   |
   = note: required by `std::convert::From::from`
```

--- a/src/librustc/traits/error_reporting.rs
+++ b/src/librustc/traits/error_reporting.rs
@@ -643,13 +643,16 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
                             .map(|s| &s == "?")
                             .unwrap_or(false);
                         let is_from = format!("{}", trait_ref).starts_with("std::convert::From<");
-                        let message = if is_try && is_from {
-                            Some(format!(
+                        let (message, note) = if is_try && is_from {
+                            (Some(format!(
                                 "`?` couldn't convert the error to `{}`",
                                 trait_ref.self_ty(),
+                            )), Some(
+                                "the question mark operation (`?`) implicitly performs a \
+                                 conversion on the error value using the `From` trait".to_owned()
                             ))
                         } else {
-                            message
+                            (message, note)
                         };
 
                         let mut err = struct_span_err!(

--- a/src/test/ui/issues/issue-32709.stderr
+++ b/src/test/ui/issues/issue-32709.stderr
@@ -4,6 +4,7 @@ error[E0277]: `?` couldn't convert the error to `()`
 LL |     Err(5)?;
    |           ^ the trait `std::convert::From<{integer}>` is not implemented for `()`
    |
+   = note: the question mark operation (`?`) implicitly performs a conversion on the error value using the `From` trait
    = note: required by `std::convert::From::from`
 
 error: aborting due to previous error

--- a/src/test/ui/try-block/try-block-bad-type.stderr
+++ b/src/test/ui/try-block/try-block-bad-type.stderr
@@ -4,6 +4,7 @@ error[E0277]: `?` couldn't convert the error to `i32`
 LL |         Err("")?;
    |                ^ the trait `std::convert::From<&str>` is not implemented for `i32`
    |
+   = note: the question mark operation (`?`) implicitly performs a conversion on the error value using the `From` trait
    = help: the following implementations were found:
              <i32 as std::convert::From<bool>>
              <i32 as std::convert::From<i16>>

--- a/src/test/ui/try-on-option.stderr
+++ b/src/test/ui/try-on-option.stderr
@@ -4,6 +4,7 @@ error[E0277]: `?` couldn't convert the error to `()`
 LL |     x?;
    |      ^ the trait `std::convert::From<std::option::NoneError>` is not implemented for `()`
    |
+   = note: the question mark operation (`?`) implicitly performs a conversion on the error value using the `From` trait
    = note: required by `std::convert::From::from`
 
 error[E0277]: the `?` operator can only be used in a function that returns `Result` or `Option` (or another type that implements `std::ops::Try`)

